File: C++/Proxy/InternetWithParentalControls/internet.cc
#include <iostream>
#include "internet.h"

void RealInternet::connect_to(const std::string &host) const {
    std::cout << "Connecting to " << host << "\n";
}


// TODO: fix clang-tidy complaint
std::unordered_set<std::string> InternetWithParentalControls::banned_sites {"banned.com", "notforchildren.com",
                                                                            "foradults.com", "nsfw.com"};

InternetWithParentalControls::InternetWithParentalControls() : internet{std::make_unique<RealInternet>()} {}

void InternetWithParentalControls::connect_to(const std::string &host) const {

    if (banned_sites.find(host) != banned_sites.end()) {
        std::cout << "Access denied...\n";
    } else {
        internet->connect_to(host);
    }
}

File: Python/Proxy/InternetWithParentalControls/internet.py
import abc


class Internet(abc.ABC):
    @abc.abstractmethod
    def connect_to(self, host: str) -> None:
        pass


class RealInternet(Internet):
    def connect_to(self, host: str) -> None:
        print(f'Connecting to {host}...')


class InternetWithParentalControls(Internet):
    banned_sites = {"banned.com", "notforchildren.com", "foradults.com", "nsfw.com"}

    def __init__(self):
        self.internet = RealInternet()

    def connect_to(self, host: str) -> None:
        if host.lower() in InternetWithParentalControls.banned_sites:
            print('Access denied...')
        else:
            self.internet.connect_to(host)


if __name__ == '__main__':
    internet = InternetWithParentalControls()
    internet.connect_to("eric.com")
    internet.connect_to("google.com")
    internet.connect_to("nsfw.com")
    internet.connect_to("notforchildren.com")
    internet.connect_to("foradults.com")
    internet.connect_to("banned.com")
    internet.connect_to("facebook.com")
